pkgs/development/beam-modules/build-mix.nix

```diff
@@ -1,100 +1,69 @@
-{ stdenv, writeText, elixir, erlang, hex, lib }:
-
-{ name
-, version
-, src
-, setupHook ? null
-, buildInputs ? []
-, beamDeps ? []
-, postPatch ? ""
-, compilePorts ? false
-, installPhase ? null
-, buildPhase ? null
-, configurePhase ? null
-, meta ? {}
-, enableDebugInfo ? false
-, ... }@attrs:
-
-with lib;
+{ stdenv, lib, elixir, erlang, hex, rebar, rebar3, fetchMixDeps }:
 
-let
+{ name, version, src, nativeBuildInputs ? [ ], meta ? { }
+, enableDebugInfo ? false, depsSha256, mixEnv ? "prod", ... }@attrs:
 
-  debugInfoFlag = lib.optionalString (enableDebugInfo || elixir.debugInfo) "--debug-info";
+let
 
-  shell = drv: stdenv.mkDerivation {
-          name = "interactive-shell-${drv.name}";
-          buildInputs = [ drv ];
+  shell = drv:
+    stdenv.mkDerivation {
+      name = "interactive-shell-${drv.name}";
+      buildInputs = [ drv ];
     };
 
-  bootstrapper = ./mix-bootstrap;
+  mixDeps = fetchMixDeps {
+    inherit src name mixEnv version;
+    sha256 = depsSha256;
+  };
 
-  pkg = self: stdenv.mkDerivation ( attrs // {
-    name = "${name}-${version}";
-    inherit version;
+  pkg = self:
+    stdenv.mkDerivation (attrs // {
+      name = "${name}-${version}";
+      inherit version;
 
-    dontStrip = true;
+      dontStrip = true;
 
-    inherit src;
+      inherit src;
 
-    setupHook = if setupHook == null
-    then writeText "setupHook.sh" ''
-       addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
-    ''
-    else setupHook;
+      nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir ];
 
-    inherit buildInputs;
-    propagatedBuildInputs = [ hex elixir ] ++ beamDeps;
+      MIX_ENV = mixEnv;
+      MIX_DEBUG = if enableDebugInfo then 1 else 0;
+      HEX_OFFLINE = 1;
+      DEBUG = if enableDebugInfo then 1 else 0; # for Rebar3 compilation
 
-    configurePhase = if configurePhase == null
-    then ''
-      runHook preConfigure
-      ${erlang}/bin/escript ${bootstrapper}
-      runHook postConfigure
-    ''
-    else configurePhase ;
+      postUnpack = ''
+        export HEX_HOME="$TMPDIR/hex"
+        export MIX_HOME="$TMPDIR/mix"
+        export MIX_DEPS_PATH="$TMPDIR/deps"
 
+        # Rebar
+        mix local.rebar rebar "${rebar}/bin/rebar"
+        mix local.rebar rebar3 "${rebar3}/bin/rebar3"
```

Comment:
I am not sure, but I think it will copy `rebar` and `rebar3` to Mix paths. Maybe it would be better to use `MIX_REBAR` and `MIX_REBAR3` to use the binaries without copying them?

Reply:
You're probably right, it does say `creating` when going through these.
Thanks for the feedback!

Reply:
For reference, it is the case, here is the code
 https://github.com/elixir-lang/elixir/blob/v1.11.3/lib/mix/lib/mix/tasks/local.rebar.ex#L68